// @emdgroup-liquid/liquid/dist/react.js
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
import { createReactComponent } from './react-component-lib';
import { defineCustomElements } from '../dist/loader';
defineCustomElements();
export const LdAccordion = /*@__PURE__*/ createReactComponent('ld-accordion');
export const LdAccordionPanel = /*@__PURE__*/ createReactComponent('ld-accordion-panel');
export const LdAccordionSection = /*@__PURE__*/ createReactComponent('ld-accordion-section');
export const LdAccordionToggle = /*@__PURE__*/ createReactComponent('ld-accordion-toggle');
export const LdBgCells = /*@__PURE__*/ createReactComponent('ld-bg-cells');
export const LdButton = /*@__PURE__*/ createReactComponent('ld-button');
export const LdCheckbox = /*@__PURE__*/ createReactComponent('ld-checkbox');
export const LdHeader = /*@__PURE__*/ createReactComponent('ld-header');
export const LdIcon = /*@__PURE__*/ createReactComponent('ld-icon');
export const LdInput = /*@__PURE__*/ createReactComponent('ld-input');
export const LdInputMessage = /*@__PURE__*/ createReactComponent('ld-input-message');
export const LdLabel = /*@__PURE__*/ createReactComponent('ld-label');
export const LdLoading = /*@__PURE__*/ createReactComponent('ld-loading');
export const LdNotice = /*@__PURE__*/ createReactComponent('ld-notice');
export const LdNotification = /*@__PURE__*/ createReactComponent('ld-notification');
export const LdOption = /*@__PURE__*/ createReactComponent('ld-option');
export const LdPagination = /*@__PURE__*/ createReactComponent('ld-pagination');
export const LdRadio = /*@__PURE__*/ createReactComponent('ld-radio');
export const LdSelect = /*@__PURE__*/ createReactComponent('ld-select');
export const LdSidenav = /*@__PURE__*/ createReactComponent('ld-sidenav');
export const LdSidenavBack = /*@__PURE__*/ createReactComponent('ld-sidenav-back');
export const LdSidenavHeading = /*@__PURE__*/ createReactComponent('ld-sidenav-heading');
export const LdSidenavNavitem = /*@__PURE__*/ createReactComponent('ld-sidenav-navitem');
export const LdSidenavSeparator = /*@__PURE__*/ createReactComponent('ld-sidenav-separator');
export const LdSidenavSlider = /*@__PURE__*/ createReactComponent('ld-sidenav-slider');
export const LdSidenavSubnav = /*@__PURE__*/ createReactComponent('ld-sidenav-subnav');
export const LdSrLive = /*@__PURE__*/ createReactComponent('ld-sr-live');
export const LdSrOnly = /*@__PURE__*/ createReactComponent('ld-sr-only');
export const LdTab = /*@__PURE__*/ createReactComponent('ld-tab');
export const LdTablist = /*@__PURE__*/ createReactComponent('ld-tablist');
export const LdTabpanel = /*@__PURE__*/ createReactComponent('ld-tabpanel');
export const LdTabpanellist = /*@__PURE__*/ createReactComponent('ld-tabpanellist');
export const LdTabs = /*@__PURE__*/ createReactComponent('ld-tabs');
export const LdToggle = /*@__PURE__*/ createReactComponent('ld-toggle');
export const LdTooltip = /*@__PURE__*/ createReactComponent('ld-tooltip');
export const LdTypo = /*@__PURE__*/ createReactComponent('ld-typo');
