// @emdgroup-liquid/liquid/dist/cjs/ld-accordion-section.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-0d41e2c9.js');
const getClassNames = require('./getClassNames-dd3906b3.js');

const ldAccordionSectionCss = ":host{display:block}";

let LdAccordionSection = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.ldaccordionchange = index.createEvent(this, "ldaccordionchange", 7);
    this.initialized = false;
  }
  updateExpandedState(newExpanded) {
    Array.from(this.el.children).forEach((child) => {
      child.setExpanded(newExpanded);
    });
    if (this.initialized) {
      this.ldaccordionchange.emit(newExpanded);
    }
  }
  handleToggleClick(ev) {
    if (ev.target.closest('ld-accordion-section') !== this.el) {
      return;
    }
    this.expanded = !this.expanded;
  }
  componentWillLoad() {
    this.updateExpandedState(this.expanded);
    this.initialized = true;
  }
  render() {
    const cl = getClassNames.getClassNames([
      'ld-accordion-section',
      this.expanded && 'ld-accordion-section--expanded',
    ]);
    return (index.h(index.Host, { class: cl, onLdaccordiontoggleclick: this.handleToggleClick.bind(this) }, index.h("slot", null)));
  }
  get el() { return index.getElement(this); }
  static get watchers() { return {
    "expanded": ["updateExpandedState"]
  }; }
};
LdAccordionSection.style = ldAccordionSectionCss;

exports.ld_accordion_section = LdAccordionSection;
