// @emdgroup-liquid/liquid/dist/components/ld-accordion-section.js
import { HTMLElement, createEvent, h, Host, proxyCustomElement } from '@stencil/core/internal/client';
import { g as getClassNames } from './getClassNames.js';

const ldAccordionSectionCss = ":host{display:block}";

let LdAccordionSection$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.ldaccordionchange = createEvent(this, "ldaccordionchange", 7);
    this.initialized = false;
  }
  updateExpandedState(newExpanded) {
    Array.from(this.el.children).forEach((child) => {
      child.setExpanded(newExpanded);
    });
    if (this.initialized) {
      this.ldaccordionchange.emit(newExpanded);
    }
  }
  handleToggleClick(ev) {
    if (ev.target.closest('ld-accordion-section') !== this.el) {
      return;
    }
    this.expanded = !this.expanded;
  }
  componentWillLoad() {
    this.updateExpandedState(this.expanded);
    this.initialized = true;
  }
  render() {
    const cl = getClassNames([
      'ld-accordion-section',
      this.expanded && 'ld-accordion-section--expanded',
    ]);
    return (h(Host, { class: cl, onLdaccordiontoggleclick: this.handleToggleClick.bind(this) }, h("slot", null)));
  }
  get el() { return this; }
  static get watchers() { return {
    "expanded": ["updateExpandedState"]
  }; }
  static get style() { return ldAccordionSectionCss; }
};
LdAccordionSection$1 = /*@__PURE__*/ proxyCustomElement(LdAccordionSection$1, [1, "ld-accordion-section", {
    "expanded": [1540],
    "initialized": [32]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["ld-accordion-section"];
  components.forEach(tagName => { switch (tagName) {
    case "ld-accordion-section":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, LdAccordionSection$1);
      }
      break;
  } });
}
defineCustomElement$1();

const LdAccordionSection = LdAccordionSection$1;
const defineCustomElement = defineCustomElement$1;

export { LdAccordionSection, defineCustomElement };
