// @emdgroup-liquid/liquid/dist/components/ld-bg-cells.js
import { HTMLElement, h, Host, proxyCustomElement } from '@stencil/core/internal/client';
import { g as getClassNames } from './getClassNames.js';
import { g as getLdAssetPath } from './getLdAssetPath.js';

const ldBgCellsCss = ".ld-bg-cells,:host{--ld-bg-cells-bg-col:var(--ld-thm-primary);--ld-bg-cells-pattern-col:var(--ld-thm-secondary);background:var(--ld-bg-cells-bg-col);height:100%;position:relative;width:100%}:host(.ld-bg-cells--bioreliance),:where(.ld-bg-cells--bioreliance){--ld-bg-cells-position:top 5% right -20%;--ld-bg-cells-size:230%}:host(.ld-bg-cells--f),:where(.ld-bg-cells--f){--ld-bg-cells-position:top 20% right -50%;--ld-bg-cells-size:180%}:host(.ld-bg-cells--hexagon),:where(.ld-bg-cells--hexagon){--ld-bg-cells-position:top 56% right -30%;--ld-bg-cells-size:200%}:host(.ld-bg-cells--millipore),:where(.ld-bg-cells--millipore){--ld-bg-cells-position:top -10% right -30%;--ld-bg-cells-size:200%}:host(.ld-bg-cells--qa-x2f-qc),:where(.ld-bg-cells--qa-x2f-qc){--ld-bg-cells-position:top -30% right -34%;--ld-bg-cells-size:170%}:host(.ld-bg-cells--safc),:where(.ld-bg-cells--safc){--ld-bg-cells-position:top 47% right -15%;--ld-bg-cells-size:290%}:host(.ld-bg-cells--sigma-aldrich),:where(.ld-bg-cells--sigma-aldrich){--ld-bg-cells-position:top 51% right -19%;--ld-bg-cells-size:280%}:host(.ld-bg-cells--t),:where(.ld-bg-cells--t){--ld-bg-cells-position:top -10% right -20%;--ld-bg-cells-size:250%}:host(.ld-bg-cells--tile),:where(.ld-bg-cells--tile){--ld-bg-cells-position:top 50% right -12%;--ld-bg-cells-size:250%}.ld-bg-cells__pattern{background:var(--ld-bg-cells-pattern-col);-webkit-mask-image:var(--ld-bg-cells-image);mask-image:var(--ld-bg-cells-image);-webkit-mask-position:var(--ld-bg-cells-position);mask-position:var(--ld-bg-cells-position);-webkit-mask-repeat:no-repeat;mask-repeat:no-repeat;-webkit-mask-size:var(--ld-bg-cells-size);mask-size:var(--ld-bg-cells-size)}.ld-bg-cells__content,.ld-bg-cells__pattern{top:0;right:0;bottom:0;left:0;position:absolute}";

let LdBgCells$1 = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    /** Cells pattern */
    this.type = 'safc';
  }
  render() {
    const assetPath = getLdAssetPath(`./assets/${this.type}-cell.svg`);
    return (h(Host, { class: getClassNames(['ld-bg-cells', `ld-bg-cells--${this.type}`]) }, h("div", { class: "ld-bg-cells__pattern", part: "pattern", style: {
        '--ld-bg-cells-image': `url(${assetPath})`,
      } }), h("div", { class: "ld-bg-cells__content", part: "content" }, h("slot", null))));
  }
  static get assetsDirs() { return ["assets"]; }
  static get style() { return ldBgCellsCss; }
};
LdBgCells$1 = /*@__PURE__*/ proxyCustomElement(LdBgCells$1, [1, "ld-bg-cells", {
    "type": [1]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["ld-bg-cells"];
  components.forEach(tagName => { switch (tagName) {
    case "ld-bg-cells":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, LdBgCells$1);
      }
      break;
  } });
}
defineCustomElement$1();

const LdBgCells = LdBgCells$1;
const defineCustomElement = defineCustomElement$1;

export { LdBgCells, defineCustomElement };
