// @emdgroup-liquid/liquid/dist/components/ld-sidenav-separator2.js
import { HTMLElement, h, Host, proxyCustomElement } from '@stencil/core/internal/client';
import { g as getClassNames } from './getClassNames.js';
import { c as closest } from './closest.js';
import { t as toggleStackToTop } from './toggleStackToTop.js';

const ldSidenavSeparatorCss = ":host{margin:var(--ld-sidenav-padding-y) var(--ld-sidenav-padding-x) 0;transform-origin:left;transition:transform var(--ld-sidenav-transition-duration-collapse-expand) ease}:host hr{border:0;border-radius:var(--ld-br-full);height:var(--ld-sp-1);margin:0;transform:scaleY(1.5);transition:transform var(--ld-sidenav-transition-duration-collapse-expand) ease;background-color:var(--ld-col-neutral-100)}:host(.ld-sidenav-separator--collapsed){transform:translateX(var(--ld-sidenav-translate-x-delta)) scaleX(var(--ld-sidenav-separator-scale-x-collapsed))}:host(.ld-sidenav-separator--collapsed) hr{transform:translateY(var(--ld-slider-navitem-move-up,0)) scaleY(1.5);transition:transform var(--ld-sidenav-transition-duration-collapse-expand) var(--ld-sidenav-transition-delay-collapse-expand) ease}";

let LdSidenavSeparator = class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.computeScaleXCollapsed = () => {
      const sidenavWidth = parseFloat(window
        .getComputedStyle(this.sidenav)
        .getPropertyValue('--ld-sidenav-width'));
      const sidenavPaddingX = parseFloat(window
        .getComputedStyle(this.sidenav)
        .getPropertyValue('--ld-sidenav-padding-x'));
      const sidenavNavitemIconSize = parseFloat(window
        .getComputedStyle(this.sidenav)
        .getPropertyValue('--ld-sidenav-navitem-icon-size'));
      return sidenavNavitemIconSize / (sidenavWidth - 2 * sidenavPaddingX);
    };
  }
  handleSidenavCollapsedChange(ev) {
    if (ev.target !== this.sidenav)
      return;
    if (this.el.parentElement &&
      !['LD-SIDENAV-SLIDER', 'LD-SIDENAV-SUBNAV'].includes(this.el.parentElement.tagName)) {
      return;
    }
    this.sidenavCollapsed = ev.detail;
    if (this.sidenav.narrow) {
      toggleStackToTop(this.el, this.sidenavCollapsed);
    }
  }
  handleSidenavBreakpointChange(ev) {
    if (ev.target !== this.sidenav)
      return;
    this.sidenavClosable = ev.detail;
  }
  componentWillLoad() {
    this.sidenav = closest('ld-sidenav', this.el);
    this.scaleXCollapsed = this.computeScaleXCollapsed() || 1;
  }
  render() {
    const cl = getClassNames([
      'ld-sidenav-separator',
      this.sidenavCollapsed &&
        !this.sidenavClosable &&
        'ld-sidenav-separator--collapsed',
    ]);
    return (h(Host, { style: {
        '--ld-sidenav-separator-scale-x-collapsed': this.scaleXCollapsed.toString(),
      }, class: cl }, h("hr", { class: "ld-sidenav-separator-line", part: "hr" })));
  }
  get el() { return this; }
  static get style() { return ldSidenavSeparatorCss; }
};
LdSidenavSeparator = /*@__PURE__*/ proxyCustomElement(LdSidenavSeparator, [1, "ld-sidenav-separator", {
    "sidenavCollapsed": [32],
    "sidenavClosable": [32],
    "scaleXCollapsed": [32]
  }, [[9, "ldSidenavCollapsedChange", "handleSidenavCollapsedChange"], [9, "ldSidenavBreakpointChange", "handleSidenavBreakpointChange"]]]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["ld-sidenav-separator"];
  components.forEach(tagName => { switch (tagName) {
    case "ld-sidenav-separator":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, LdSidenavSeparator);
      }
      break;
  } });
}
defineCustomElement();

export { LdSidenavSeparator as L, defineCustomElement as d };
